### PR TITLE
Add cert-rotation e2e test

### PR DESCRIPTION
**Pull request by namnx228**
This PR adds an e2e test for the Ironic certificate rotation feature.
The Ironic certificate rotation feature introduces the use of cert-manger to generate new TLS certificates and renew these certificates when they are expired. The current integration test can ensure if ironic can adopt the new certificates generated by cert-manager, so this PR aims to test the if certificates can be renewed, and ironic can continue to function well with the new certificates.
Test scenario:
- Manually trigger the certificate renewal by deleting the secrets that contain Ironic certificates.
- Check if ironic contains which use the certificates can restart. After restarting, they will adopt the new certificates.
- Check if the containers are in the running state.

### CODE REVIEW

File: test/e2e/cert_rotation_test.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	Eventually(func() error {
+		ironicPod, err := getPodFromDeployment(targetCluster, ironicDeployment, ironicNamespace)
+		if err != nil {
+			return err
+		}
+		if ironicPod.Status.Phase == corev1.PodRunning {
+			return nil
+		}
+
+		return errors.New("Ironic pod cannot be in running state")
```
Comment (fmuyassarov) [inline test/e2e/cert_rotation_test.go:38]:
```suggestion
		return errors.New("Ironic pod is not in running state")
```

File: test/e2e/cert_rotation_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	By("Force the cert-manager to regenerate the certificate by deleting the secrets")
+
+	secretList := []string{
+		"ironic-cert",
+		"ironic-inspector-cert",
+		"mariadb-cert",
+	}
+	for _, secretName := range secretList {
+		err := targetClusterClientSet.CoreV1().Secrets(ironicNamespace).Delete(ctx, secretName, metav1.DeleteOptions{})
+		Expect(err).To(BeNil(), "Cannot detele the this secret: %s", secretName)
```
Comment (fmuyassarov) [inline test/e2e/cert_rotation_test.go:50]:
```suggestion
		Expect(err).To(BeNil(), "Cannot delete this secret: %s", secretName)
```

File: test/e2e/cert_rotation_test.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	containerNumRestart["mariadb"] = 0
+	Expect(err).To(BeNil())
+	ironicPod, err := getPodFromDeployment(targetCluster, ironicDeployment, ironicNamespace)
+	Expect(err).To(BeNil())
+	for _, container := range ironicPod.Status.ContainerStatuses {
+		if _, exist := containerNumRestart[container.Name]; exist {
+			containerNumRestart[container.Name] = container.RestartCount
+		}
+	}
+	Logf("Wait containers which got a new certificate until they are succesfully restarted")
```
Comment (fmuyassarov) [inline test/e2e/cert_rotation_test.go:67]:
```suggestion
	Logf("Wait until containers successfully restarted after renewing certs")
```

File: test/e2e/cert_rotation_test.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+			for _, container := range ironicPod.Status.ContainerStatuses {
+				if oldNumRestart, exist := containerNumRestart[container.Name]; exist {
+					if !(oldNumRestart < container.RestartCount) {
+						return fmt.Errorf("%s is not restarted", container.Name)
+					}
+				}
+			}
+			return nil
+		}
+		return errors.New("Ironic pod cannot be in running state")
```
Comment (fmuyassarov) [inline test/e2e/cert_rotation_test.go:84]:
```suggestion
		return errors.New("Ironic is not in running state")
```

File: test/e2e/cert_rotation_test.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+		ironicPod, err := getPodFromDeployment(targetCluster, ironicDeployment, ironicNamespace)
+		if err != nil {
+			return err
+		}
+
+		if ironicPod.Status.Phase == corev1.PodRunning {
+			for _, container := range ironicPod.Status.ContainerStatuses {
+				if oldNumRestart, exist := containerNumRestart[container.Name]; exist {
+					if !(oldNumRestart < container.RestartCount) {
+						return fmt.Errorf("%s is not restarted", container.Name)
```
Comment (furkatgofurov7) [inline test/e2e/cert_rotation_test.go:78]: I am not sure if I am getting the logic properly here, would not this be always true? Also, can't pod restart in other cases than after regenerating certs, i.e node unavailability, resource limitation?

**Review by furkatgofurov7 — COMMENTED** [review]
Great work, some minor suggestions/questions from my side:

File: test/e2e/cert_rotation_test.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	for _, secretName := range secretList {
+		err := targetClusterClientSet.CoreV1().Secrets(ironicNamespace).Delete(ctx, secretName, metav1.DeleteOptions{})
+		Expect(err).To(BeNil(), "Cannot detele the this secret: %s", secretName)
+	}
+
+	By("Wait for containers in the ironic pod to be restarted")
+	Logf("Get the current number of time containers were restarted")
+	containerNumRestart := make(map[string]int32)
+	containerNumRestart["ironic-api"] = 0
+	containerNumRestart["ironic-conductor"] = 0
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/cert_rotation_test.go:57]: ironic-inspector doesn't restart, but its reverse proxy does, and this proxy is running inside ironic-api.

File: test/e2e/cert_rotation_test.go (hunk 5 above)
Comment (namnx228) [reply test/e2e/cert_rotation_test.go:78]: I'm expecting that whenever the certificate is renewed, the container which is mounting it should restart itself to load the new cert into its process. That is true that something else may cause the restart. Do you have any ideas how we can improve it?

Comment (furkatgofurov7) [reply test/e2e/cert_rotation_test.go:78]: I had an idea of comparing old and newly regenerated certs and once we are sure it is a new one, check if same pod restarted, but based on the conversation, I do agree that might not be a necessary thing to do, so I am fine leaving it as is.

**Review by kashifest — APPROVED** [review]

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

**Pull request merged**